File: src/utils/colorUtils.ts
export function parseColorString(rawValue: string): number[] {
  const stripped = rawValue.replace(/[\[\]]/g, '');
  const parts    = stripped.split(/[,\s]+/).filter(Boolean);
  return parts.slice(0, 4).map(p => Math.min(255, Math.max(0, Math.round(parseFloat(p)))));
}

export function rgbToHex(rgbChannels: number[]): string {
  const [r = 0, g = 0, b = 0] = rgbChannels;
  return '#' + [r, g, b].map(v => Math.min(255, Math.max(0, v)).toString(16).padStart(2, '0')).join('');
}

export function hexToRgb(hexString: string): number[] | null {
  let normalized = hexString.trim().replace('#', '');
  if (normalized.length === 3) normalized = normalized.split('').map(c => c + c).join('');
  if (normalized.length !== 6) return null;
  const red   = parseInt(normalized.slice(0, 2), 16);
  const green = parseInt(normalized.slice(2, 4), 16);
  const blue  = parseInt(normalized.slice(4, 6), 16);
  if (isNaN(red) || isNaN(green) || isNaN(blue)) return null;
  return [red, green, blue];
}

export function isValidHex(hexString: string): boolean {
  const normalized = hexString.trim().replace('#', '');
  return (normalized.length === 3 || normalized.length === 6) && /^[0-9a-fA-F]+$/.test(normalized);
}

export function colorDistance(rgbA: number[], rgbB: number[]): number {
  const deltaR = (rgbA[0] ?? 0) - (rgbB[0] ?? 0);
  const deltaG = (rgbA[1] ?? 0) - (rgbB[1] ?? 0);
  const deltaB = (rgbA[2] ?? 0) - (rgbB[2] ?? 0);
  return Math.sqrt(deltaR * deltaR + deltaG * deltaG + deltaB * deltaB);
}

export function colorsAreEqual(rgbA: number[], rgbB: number[]): boolean {
  return rgbA[0] === rgbB[0] && rgbA[1] === rgbB[1] && rgbA[2] === rgbB[2];
}

File: src-tauri/src/color_utils.rs
pub fn parse_color_string(raw_value: &str) -> Vec<u8> {
    raw_value
        .trim()
        .trim_matches(|c| c == '[' || c == ']')
        .split(|c: char| !c.is_ascii_digit() && c != '.')
        .filter(|part| !part.is_empty())
        .filter_map(|part| part.parse::<f64>().ok())
        .take(4)
        .map(|v| v.clamp(0.0, 255.0) as u8)
        .collect()
}

pub fn color_list_to_string(rgb_channels: &[u8]) -> String {
    let channel_strings: Vec<String> = rgb_channels.iter().map(|v| v.to_string()).collect();
    format!("[ {} ]", channel_strings.join(", "))
}

pub fn rgb_to_hex(rgb_channels: &[u8]) -> String {
    if rgb_channels.len() >= 3 {
        format!("#{:02x}{:02x}{:02x}", rgb_channels[0], rgb_channels[1], rgb_channels[2])
    } else {
        "#000000".to_string()
    }
}

pub fn color_distance(color_a: &[u8], color_b: &[u8]) -> f64 {
    if color_a.len() < 3 || color_b.len() < 3 {
        return f64::MAX;
    }
    let delta_r = color_a[0] as f64 - color_b[0] as f64;
    let delta_g = color_a[1] as f64 - color_b[1] as f64;
    let delta_b = color_a[2] as f64 - color_b[2] as f64;
    (delta_r * delta_r + delta_g * delta_g + delta_b * delta_b).sqrt()
}
